src/runtime/GlobalObjectBuiltinWeakRef.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2018-present Samsung Electronics Co., Ltd
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this library; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
+ *  USA
+ */
+
+#include "Escargot.h"
+#include "GlobalObject.h"
+#include "Context.h"
+#include "VMInstance.h"
+#include "WeakRefObject.h"
+#include "IteratorObject.h"
+#include "NativeFunctionObject.h"
+#include "ToStringRecursionPreventer.h"
+
+namespace Escargot {
+
+Value builtinWeakRefConstructor(ExecutionState& state, Value thisValue, size_t argc, Value* argv, Optional<Object*> newTarget)
+{
+    if (!newTarget.hasValue()) {
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, ErrorObject::Messages::GlobalObject_ConstructorRequiresNew);
+    }
+    if (argc == 0 || !argv[0].isObject()) {
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, "target is not object");
+    }
+
+    // Let weakRef be ? OrdinaryCreateFromConstructor(NewTarget, "%WeakRefPrototype%", « [[WeakRefTarget]] »).
+    Object* proto = Object::getPrototypeFromConstructor(state, newTarget.value(), [](ExecutionState& state, Context* constructorRealm) -> Object* {
+        return constructorRealm->globalObject()->weakRefPrototype();
+    });
+    WeakRefObject* weakRef = new WeakRefObject(state, proto, argv[0].asObject());
+
+    return weakRef;
+}
+
+
+Value builtinWeakRefDeRef(ExecutionState& state, Value thisValue, size_t argc, Value* argv, Optional<Object*> newTarget)
+{
+    if (!thisValue.isObject() || !thisValue.asObject()->isWeakRefObject()) {
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, ErrorObject::Messages::GlobalObject_CalledOnIncompatibleReceiver);
+    }
+    // Let weakRef be the this value.
+    WeakRefObject* weakRef = thisValue.asObject()->asWeakRefObject();
+    // If target is not empty
+    if (weakRef->getTarget()) {
+        return weakRef->getTarget();
+    }
```

Comment:
IMO, its better to check the target value and return the correct value (undefined or target value) inside `WeakRefObject::getTarget()` instead of repetitively calling `getTarget` function here (similar to `WeakMapObject::get()`).